magento/pwa-studio
# Support for express server compression middleware.

## Description

The `express` server that serves the app provides a middleware that applies GZIP compression if not applied already on the responses requested by the storefront. Compression is a very important but sensitive feature to manage because it can be applied at different layers of a system. One can apply it at the proxy layer or application layer and overdoing it can cause performance degradation. Hence the express server compression will be controlled by an environment variable. If someone wants to apply compression at the express layer, they have to set the value of `ENABLE_EXPRESS_SERVER_COMPRESSION` to `true`. By default, it will be set to `false`.

`ENABLE_EXPRESS_SERVER_COMPRESSION` in a sample project's `.env` file created using the `create-pwa` scafolding tool:

![image](https://user-images.githubusercontent.com/35203638/106299245-8f6eb100-621a-11eb-94e9-fc06236f6f1c.png)


## Related Issue
Closes PWA-1108.

## Acceptance
Should apply GZIP compression on responses if `ENABLE_EXPRESS_SERVER_COMPRESSION` to `true`.

### Verification Stakeholders
@jimbo 
@dpatil-magento 

### Verification Steps
1. Change/create env variable in the .env file `ENABLE_EXPRESS_SERVER_COMPRESSION=true`
2. Run the app. Check both `yarn watch:venia` and `yarn build && yarn stage:venia`
3. Open the app and check the network tab for responses **with** a header `Content-Encoding: gzip`.
4. Change the value of `ENABLE_EXPRESS_SERVER_COMPRESSION` to `false`.
5. Restart the app.
6. Open the app and check the network tab for responses **without** a header `Content-Encoding: gzip`.

## Screenshots / Screen Captures (if appropriate)

When `ENABLE_EXPRESS_SERVER_COMPRESSION=true`

![image](https://user-images.githubusercontent.com/35203638/106303778-25f1a100-6220-11eb-8fa7-dc711f2a50c3.png)

When `ENABLE_EXPRESS_SERVER_COMPRESSION=false`

![image](https://user-images.githubusercontent.com/35203638/106304025-72d57780-6220-11eb-81b0-499ae6f40801.png)

## Checklist

* I have added tests to cover my changes, if necessary.
* I have added translations for new strings, if necessary.
* I have updated the documentation accordingly, if necessary.
